### PR TITLE
Arbitrary function call in UDF

### PR DESCRIPTION
**Pull request by wjxiz1992**
Enable arbitrary function call in UDF

Signed-off-by: Allen Xu <allxu@nvidia.com>

### CODE REVIEW

**Comment by wjxiz1992** [comment]
@seanprime7 Added initial test case for this feature.

**Comment by firestarman** [comment]
Better to create an issue for this too.

File: udf-compiler/src/test/scala/com/nvidia/spark/OpcodeSuite.scala [hunk 1]
```diff
@@ -35,6 +35,7 @@ class OpcodeSuite extends FunSuite {
 
   val conf: SparkConf = new SparkConf()
     .set("spark.sql.extensions", "com.nvidia.spark.udf.Plugin")
+    .set("spark.rapids.sql.test.enabled", "true")
```
Comment (seanprime7) [inline udf-compiler/src/test/scala/com/nvidia/spark/OpcodeSuite.scala:38]: @wjxiz1992 Is this needed?  It makes all the tests that are supposed to fall back to CPU fail.

Comment (wjxiz1992) [reply udf-compiler/src/test/scala/com/nvidia/spark/OpcodeSuite.scala:38]: I will remove this after the feature is implemented. It's enabled now to see the actual error message like `Unsupported instruction: 182`.

Comment (seanprime7) [reply udf-compiler/src/test/scala/com/nvidia/spark/OpcodeSuite.scala:38]: I see.  Okay.

**Comment by seanprime7** [comment]
@wjxiz1992 @nvnedelman do you guys see if there is any other method call pattern that needs to be added to the test suite?

**Comment by seanprime7** [comment]
Resolved merge conflicts.

File: udf-compiler/src/main/scala/com/nvidia/spark/udf/State.scala [hunk 2]
```diff
@@ -117,23 +118,23 @@ case class State(locals: Array[Expression],
...
+                        children: Seq[Expression],
+                        objref: Option[Expression]): State = {
     val max = lambdaReflection.maxLocals
-    val params: Seq[(CtClass, Expression)] = lambdaReflection.parameters.view.zip(children)
-    val (locals, _) = params.foldLeft((new Array[Expression](max), 0)) { (l, p) =>
-      val (locals: Array[Expression], index: Int) = l
-      val (param: CtClass, argExp: Expression) = p
-
-      val newIndex = if (param == CtClass.doubleType || param == CtClass.longType) {
+    val _children = lambdaReflection.capturedArgs ++ children
```
Comment (abellina) [inline udf-compiler/src/main/scala/com/nvidia/spark/udf/State.scala:124]: nit, we don't tend to use underscore in variable names. Could call this `capturedAndChildren`?

File: udf-compiler/src/test/scala/com/nvidia/spark/OpcodeSuite.scala [hunk 3]
```diff
@@ -2086,4 +2086,202 @@ class OpcodeSuite extends FunSuite {
     val ref = dataset.select(lit(Array.empty[String]).as("emptyArrOfStr"))
     checkEquiv(result, ref)
   }
+
+  test("final static method call inside UDF - ") {
```
Comment (abellina) [inline udf-compiler/src/test/scala/com/nvidia/spark/OpcodeSuite.scala:2090]: minor nit, there's a ` - ` at the end of some of these tests. Is that on purpose?

**Review by abellina — COMMENTED** [review]
Just have a couple of nits @seanprime7. Looks good to me though.

File: udf-compiler/src/main/scala/com/nvidia/spark/udf/State.scala (hunk 2 above)
Comment (seanprime7) [reply udf-compiler/src/main/scala/com/nvidia/spark/udf/State.scala]: Renamed it to `args`

File: udf-compiler/src/test/scala/com/nvidia/spark/OpcodeSuite.scala (hunk 3 above)
Comment (seanprime7) [reply udf-compiler/src/test/scala/com/nvidia/spark/OpcodeSuite.scala]: Removed the trailing `-`

**Comment by abellina** [comment]
build

**Review by abellina — APPROVED** [review]

**Pull request merged**